### ISSUE TITLE
Does the prompt cache work?   I got an alignment assert when I turned it on.

### ISSUE THREAD

**Issue by johnburkey** [cmp-nct/ggllm.cpp#42]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [Y ] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [Y ] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [Y ] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [Y ] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

**Comment by cmp-nct** [comment]
It's untested, I'll see to it
Most likely not a big thing

**Comment by johnburkey** [comment]
Thanks I got around the alignment assert but the restored context looked off - happily returning an unrelated answer - guessing you know how to debug that 1000x  faster than me .

For the alignment I just added aligned allocs and noticed the size of func was returning size of element instead of size of buf so just set it to the alloced size.

Let me know if you need anything from me.

**Comment by johnburkey** [comment]
And I have some statistics buddies if you are still diving on k clusters - who knows they may have cool ideas

**Comment by cmp-nct** [comment]
Feel free to push a PR with your fix :)

Regarding work on k-quants: Those are always very time intense as they need to work with cuda implementations (optimally also with metal and opencl), need to be integrated into a ton of code and maintained. So it's an area where any change can cause a chain of changes.
Involvement is always welcome, there is a lot that can be improved and optimized.

**Comment by johnburkey** [comment]
ill see what I can do about the alignment fix, but hiding behind that was the context didn't seem to be restored correctly (it replied to a different question than asked. ) Ill check it again and report back.

**Comment by johnburkey** [comment]
Yeah there seems to be enough bugs in there I don't think I should check in.   Its crashing with a buffer overrun in ggml_compute_forward_dup_f32 during context saving for this model : falcon7b-instruct.ggmlv3.q4_0.bin.

**Comment by johnburkey** [comment]
If you get that context saving working ill beat on it a bit- I think its interesting stuff.

**Comment by cmp-nct** [comment]
> Yeah there seems to be enough bugs in there I don't think I should check in. It's crashing with a buffer overrun in ggml_compute_forward_dup_f32 during context saving for this model : falcon7b-instruct.ggmlv3.q4_0.bin. (from https://huggingface.co/TheBloke/falcon-7b-instruct-GGML)

I've never seen that happening (in thousands of tests). The v3 version is quite outdated but should work.
Can you provide the full command line and output log of it happening?

**Comment by johnburkey** [comment]
ok, pulled fresh, same issues- Im thinking maybe its because im building on the Mac without DLLAMA_CUBLAS - because the Mac has its own BLAS (Accelerate).   It builds fine, but im guessing is executing a few of the ifdefs flipped different ways?   Is there anyway for you to run that way (without CUBLAS?) and see if you get the crash?   Or would you like me to continue investigating in some way?

I ran it with these args --prompt-cache promptCache -t 8 -ngl 100 -b 1 -m ..../falcon7b-instruct.ggmlv3.q4_0.bin -p "What is a falcon?\n### Response:"

It crashes immediately typically in that same routine (ggml_compute_forward_dup)

**Comment by johnburkey** [comment]
from ggml_compute_forward_dup_f32()

````
```
  if (src0->type == dst->type &&
        ne00 == ne0 &&
        nb00 == GGML_TYPE_SIZE[src0->type] && nb0 == GGML_TYPE_SIZE[dst->type]) {
        // copy by rows
        const size_t rs = ne00*nb00;
        for (int64_t i03 = 0; i03 < ne03; i03++) {
            for (int64_t i02 = 0; i02 < ne02; i02++) {
                for (int64_t i01 = ir0; i01 < ir1; i01++) {

// its this memcpy
// its the src ptr thats bad (or in any case not big enough for the cpy)

                    memcpy(
                        ((char *)  dst->data + i01*nb1  + i02*nb2  + i03*nb3),
                        ((char *) src0->data + i01*nb01 + i02*nb02 + i03*nb03),
                        rs);
                }
            }
        }
        return;
    }
```
````

**Comment by cmp-nct** [comment]
I'll take a deeper look at it.
From first glance: It can not work at the moment, it's not made for multiquery attention.
The function converts the kv cache to 3d, stores it that way. Which I don't understand given that kv is internally a 1d tens.

Just for my understanding, I've not used the sessions before: What is the main purpose for you ?
Are you processing a long input and wish to work with that with multiple "post-prompts" ?

**Comment by johnburkey** [comment]
you know the fact that the parameters for the loop come from source implies that the src tensor might just be misconfigured, and that its not a type mismatch but just the tensor misreporting during the llama_save_session_file call of llama_copy_state_data

**Comment by cmp-nct** [comment]
No, such problems always look like that. It's a heap buffer overflow which destroys the tensor content, so the assertion breaks given the tensor looks stupid now.

I've just pushed a bugfix which appears to work. I tested it with a 1kb file and it loaded and stored correctly.
The session functionality is not fully supported yet with the new features. so using "-enc" or the system prompt will break it.

**Comment by johnburkey** [comment]
sweet will test

**Comment by cmp-nct** [comment]
-enc and -sys should also work with it by now

**Comment by johnburkey** [comment]
works so far, testing enc/sys don't know what they do, so will read about that first :-)
